src/Emogrifier.php

```diff
@@ -1267,24 +1282,70 @@ private function splitCssAndMediaQuery($css)
             $mediaTypesExpression = '|' . implode('|', array_keys($this->allowedMediaTypes));
         }
 
-        $media = '';
-        $cssForAllowedMediaTypes = preg_replace_callback(
-            '#@media\\s+(?:only\\s)?(?:[\\s{\\(]\\s*' . $mediaTypesExpression . ')\\s*[^{]*+{.*}\\s*}\\s*#misU',
-            function ($matches) use (&$media) {
-                $media .= $matches[0];
-            },
-            $cssWithoutComments
+        $cssSplitForAllowedMediaTypes = preg_split(
+            '#(@media\\s+(?:only\\s)?(?:[\\s{\\(]\\s*' . $mediaTypesExpression . ')\\s*[^{]*+{.*}\\s*}\\s*)#misU',
+            $cssWithoutComments,
+            -1,
+            PREG_SPLIT_DELIM_CAPTURE
         );
 
-        // filter the CSS
-        $search = [
+        // filter the CSS outside/between allowed @media rules
+        $cssCleaningMatchers = [
             'import/charset directives' => '/^\\s*@(?:import|charset)\\s[^;]+;/misU',
             'remaining media enclosures' => '/^\\s*@media\\s[^{]+{(.*)}\\s*}\\s/misU',
         ];
 
-        $cleanedCss = preg_replace($search, '', $cssForAllowedMediaTypes);
+        $splitCss = [];
+        foreach ($cssSplitForAllowedMediaTypes as $splitPartIndexWhereOddIsMediaRule => $cssPart) {
+            if ($splitPartIndexWhereOddIsMediaRule % 2 !== 0) {
+                // @media rule
+                preg_match('/^([^{]*+){(.*)}[^}]*+$/s', $cssPart, $matches);
+                $media = $matches[1];
+                $css = $matches[2];
+            } else {
+                // CSS outside/between allowed @media rules
+                $media = '';
+                $css = preg_replace($cssCleaningMatchers, '', $cssPart);
+            }
+            $splitCss[] = compact('css', 'media');
```

Comment:
I propose setting `$splitCss[]` directly in the `if/else`. This will remove another layer of indirection, i.e., will make the code more straightforward to read.

Reply:
Changed as suggested, and renamed `$css` -> `$cleanedCss` in the non-media branch so that the original parameter `$css` is not overwritten (and thus remains available for debugging etc).